### PR TITLE
781 wys declarative partitioning

### LINKED ISSUE #781: Change WYS inheretence -> declarative partitioning
Update `wys.mobile_sign_installations` and `wys.raw_data` to declarative partitioning. 
- `wys.raw_data`:
  - yearly tables are approaching 30GB+30GB indices. Further partition into monthly tables. 
  - Add new functions for partition creates + new airflow task check_partitions (similar to recent Miovision addition)
- `wys.mobile_sign_installations`:
  - 25 ward tables is very cluttered! Change to declarative partitioning. No other changes required. 
  - Intersections_tmc requires minor update - change inserts to target parent table.

### PR DESCRIPTION
**Pull request by gabrielwol**
## What this pull request accomplishes:

- [x] Adjust `wys.raw_data` to use declarative partitioning 
- [x] Adjust `wys.mobile_sign_installations` to use declarative partitioning (in comment section)
- added two tasks to `dags/common_tasks.py`: `check_jan_1st`, `check_1st_of_month` which are used to determine if new partitions should be created. 
  - Adjusted 3 existing DAGs `pull_miovision`, `vds_*` to use these common_tasks. 
  - Added partition creates to `pull_wys` DAG using these common tasks.
- Removed outdated inheritence partition creates/files: 
  - Remove wys from `eoy_table_create` DAG
  - Delete: `wys_eoy_create_table.py`, `create-table-raw_data_children.sql`, `function_create_raw_table.sql`, `wys/api/sql/insert_trigger.sql`.
- Adds two functions: `wys.create_mm_nested_raw_data_partitions`, `wys.create_yyyy_raw_data_partition` which are copied from recent miovision partitioning (#735 )
- Adjustments to `wys/api/python/wys_api.py` and `wys_google_sheet.py` to insert into parent table and not partition (declarative partitioning handles this).

## Issue(s) this solves:
- Closes #781 

## What, in particular, needs to reviewed:
- Not much. The `wys.raw_data` changes have already been implemented. The `wys.mobile_sign_installations` changes in the comment section below will have to be made by an open_data admin. 

## What needs to be done by a sysadmin after this PR is merged
- Check the 3 DAGs parse before merging.

### CODE REVIEW

**Comment by gabrielwol** [comment]
Here's what's remaining to run for `wys.mobile_sign_installations`:

- [ ] drop the dependencies! (VIEW open_data.wys_mobile_detailed, MATERIALIZED VIEW wys.mobile_api_id)
```sql
SELECT public.deps_save_and_drop_dependencies(
    'wys' COLLATE pg_catalog."C",
    'mobile_sign_installations' COLLATE pg_catalog."C"
);
ALTER TABLE wys.mobile_sign_installations_staging RENAME TO mobile_sign_installations;
```
- [ ] drop ward tables inheretence and remove pkey (to be replaced with one containing ward_no)
```sql
DO $do$
DECLARE
	ward INTEGER;
    ward_table TEXT;
    pkey TEXT;
BEGIN
	FOR ward IN 1..25 LOOP
        ward_table := 'ward_'||ward::text;
        pkey := 'ward_'||ward::text||'_pkey';
        EXECUTE FORMAT($$
                       ALTER TABLE wys.%s NO INHERIT wys.mobile_sign_installations;
                       ALTER TABLE wys.%s DROP CONSTRAINT IF EXISTS %s;
                       $$, ward_table, ward_table, pkey);
	END LOOP;
END;
$do$ LANGUAGE plpgsql
```
- [ ] Drop existing table and replace with partitioned version. 
```sql
DROP TABLE wys.mobile_sign_installations; 
ALTER TABLE wys.mobile_sign_installations_staging RENAME TO mobile_sign_installations;
```

- [ ] Attach all the partitions!
```sql
DO $do$
DECLARE
	ward INTEGER;
BEGIN
	FOR ward IN 1..25 LOOP
        ward_table := 'ward_'||ward::text;
        EXECUTE FORMAT($$ALTER TABLE wys.mobile_sign_installations
                         ATTACH PARTITION wys.%s FOR VALUES IN (ward::integer);
                       $$, ward_table);
	END LOOP;
END;
$do$ LANGUAGE plpgsql
```

- [ ] Restore dependencies 
```sql
SELECT public.deps_restore_dependencies('wys', 'mobile_sign_installations');
```

**Comment by gabrielwol** [comment]
The column order in github / bigdata must have been different because when I recreated the table with the github definition, the [insert is failing](https://trans-bdit.intra.prod-toronto.ca/airflow/dags/pull_wys/grid?dag_run_id=scheduled__2023-11-29T20%3A00%3A00%2B00%3A00&task_id=pull_wys&tab=logs). I added the column order to the insert in the latest commit, but `pull_wys.pull_wys` is going to keep failing until we merge this.

File: dags/eoy_create_tables.py [hunk 1]
```diff
@@ -136,18 +128,6 @@ def congestion_create_table(dt):
...
-                                    dag = dag,
-                                    op_kwargs = {'pg_hook': wys_bot,
-                                                 'dt': '{{ ds }}'}
-                                    )
-wys_replace_trigger = PythonOperator(task_id='wys_replace_trigger',
-                                    python_callable = replace_wys_raw_data_trigger,
-                                    dag = dag,
-                                    op_kwargs = {'pg_hook': wys_bot,
-                                                 'dt': '{{ ds }}'})
-
```
Comment (tahaislam) [inline dags/eoy_create_tables.py:150]: This [line](https://github.com/CityofToronto/bdit_data-sources/blob/bdb853f4fb869c15e795aa9e906cfba46bb1d6ef/dags/eoy_create_tables.py#L146) needs to be deleted as well. Also check if any of these deleted variables is used anywhere in the file.

File: dags/common_tasks.py [hunk 2]
```diff
@@ -79,3 +79,19 @@ def copy_table(conn_id:str, table:Tuple[str, str]) -> None:
...
+def check_jan_1st(ds=None): #check if Jan 1 to trigger partition creates. 
+    from datetime import strptime
+    start_date = strptime(ds, '%Y-%m-%d')
+    if start_date.month == 1 and start_date.day == 1:
+        return True
+    return False
+
+@task.short_circuit(ignore_downstream_trigger_rules=False) #only skip immediately downstream task
+def check_1st_of_month(ds=None): #check if 1st of Month to trigger partition creates. 
+    from datetime import strptime
```
Comment (tahaislam) [inline dags/common_tasks.py:93]: same as above

File: dags/common_tasks.py [hunk 3]
```diff
@@ -79,3 +79,19 @@ def copy_table(conn_id:str, table:Tuple[str, str]) -> None:
         cur.execute(insert_query)
     
     LOGGER.info(f"Successfully copied {table[0]} to {table[1]}.")
+
+@task.short_circuit(ignore_downstream_trigger_rules=False) #only skip immediately downstream task
+def check_jan_1st(ds=None): #check if Jan 1 to trigger partition creates. 
+    from datetime import strptime
```
Comment (tahaislam) [inline dags/common_tasks.py:85]: I think `strptime` is under `datetime.datetime`

**Review by tahaislam — CHANGES_REQUESTED** [review]
Good work, the code is looking simpler. I left some comments for you to fix. Then, I'll review/test the rest of the files.
Also, I think it should have been better to keep this PR only for WYS code and all other changes to the other DAGs to be included in a separate DAG to avoid any mass DAG failure and also to make it easier to get reviewed and tested. Maybe this's something to consider in future PRs

File: dags/common_tasks.py [hunk 4]
```diff
@@ -79,3 +79,19 @@ def copy_table(conn_id:str, table:Tuple[str, str]) -> None:
         cur.execute(insert_query)
     
     LOGGER.info(f"Successfully copied {table[0]} to {table[1]}.")
+
+@task.short_circuit(ignore_downstream_trigger_rules=False) #only skip immediately downstream task
+def check_jan_1st(ds=None): #check if Jan 1 to trigger partition creates.
```
[the remark this answers is not in the document]
Comment (gabrielwol) [reply dags/common_tasks.py:84]: Good idea.

**Comment by gabrielwol** [comment]
There must be a difference in the definition between `mobile_sign_installations` and `ward_*` tables which is causing [this insert](https://trans-bdit.intra.prod-toronto.ca/airflow/dags/pull_wys/grid?dag_run_id=scheduled__2023-12-04T20%3A00%3A00%2B00%3A00&task_id=read_google_sheets&tab=logs) to fail now that the insert is being directed to the parent table and not the child table. Need to return to this before finalizing this PR.

**Comment by gabrielwol** [comment]
All the tasks in the latest runs of [pull_wys](https://10.162.65.124/airflow/dags/pull_wys/grid) and [wys_monthly_summary](https://10.162.65.124/airflow/dags/wys_monthly_summary/grid?base_date=2023-11-02T23%3A59%3A59.999Z&dag_run_id=scheduled__2023-11-02T07%3A00%3A00%2B00%3A00&tab=graph&task_id=wys_mobile_summary) DAGs are now running successfully.

**Review by tahaislam — APPROVED** [review]
Great work

**Pull request merged**